[SaferC++] Deploy more smart pointers in BackForwardCache

AlexsanderDamaceno opened the pull request:
#### 9d3d5c8eac51d9b1ad84b19d2207d9352ca1c574
<pre>
[SaferC++] Deploy more smart pointers in BackForwardCache
<a href="https://bugs.webkit.org/show_bug.cgi?id=301742">https://bugs.webkit.org/show_bug.cgi?id=301742</a>

Reviewed by Anne van Kesteren.

[SaferC++] Deploy more smart pointers in BackForwardCache

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::canCacheFrame):
(WebCore::BackForwardCache::dump const):

Canonical link: <a href="https://commits.webkit.org/302603@main">https://commits.webkit.org/302603@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d975eed67d4c1ad7e82bcfec70e859ff6bc3dfc7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129533 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1790 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40372 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136917 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80965 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131404 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1722 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1666 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98665 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66527 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132480 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116025 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79313 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1256 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34156 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80190 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109726 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34654 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139390 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1580 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1578 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107186 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112366 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107028 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1287 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30879 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54272 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20227 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1651 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65014 "Found 6 new failures in history/BackForwardCache.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1471 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1505 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1573 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/history/BackForwardCache.cpp:
```diff
@@ -124,7 +124,7 @@ static bool canCacheFrame(LocalFrame& frame, DiagnosticLoggingClient& diagnostic
 
     URL currentURL = documentLoader->url();
-    URL newURL = frameLoader->provisionalDocumentLoader() ? frameLoader->provisionalDocumentLoader()->url() : URL();
+    URL newURL = frameLoader->protectedProvisionalDocumentLoader() ? frameLoader->protectedProvisionalDocumentLoader()->url() : URL();
```
You should put provisionalDocumentLoader in a local variable. It's a smell when you have to use the same protected method within the same scope.

annevk commented on Source/WebCore/history/CachedPage.h:
```diff
@@ -50,6 +50,7 @@ class CachedPage final : public CanMakeCheckedPtr<CachedPage> {
     void clear();
 
     Page& page() const { return m_page.get(); }
+    WeakRef<Page> protectedPage() { return m_page; }
```
Protected refers to Ref or RefPtr, not WeakPtr.

annevk commented on Source/WebCore/history/BackForwardCache.cpp:
```diff
@@ -322,8 +323,10 @@ void BackForwardCache::dump() const
-            WTFLogAlways("  Page %p, document %p %s", &(*cachedPage)->page(), (*cachedPage)->document(), (*cachedPage)->document() ? (*cachedPage)->document()->url().string().utf8().data() : "");
+        if (auto* cachedPage = std::get_if<UniqueRef<CachedPage>>(&item.value)) {
+            RefPtr document = (*cachedPage)->document();
+            WTFLogAlways("  Page %p, document %p %s", &(*cachedPage)->page(), document.get(), document.get() ? document->url().string().utf8().data() : "");
```
```suggestion
            WTFLogAlways("  Page %p, document %p %s", &(*cachedPage)->page(), document.get(), document ? document->url().string().utf8().data() : "");
```

annevk reviewed (approved):
Looks good modulo nit.

AlexsanderDamaceno committed:
[SaferC++] Deploy more smart pointers in BackForwardCache
https://bugs.webkit.org/show_bug.cgi?id=301742

Reviewed by Anne van Kesteren.

[SaferC++] Deploy more smart pointers in BackForwardCache

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::canCacheFrame):
(WebCore::BackForwardCache::dump const):

Canonical link: https://commits.webkit.org/302603@main
